File: go/internal/xrpl/serialize.go
package xrpl

import (
	"errors"
	"fmt"
)

// AccountID is the 20-byte XRPL account identifier.
type AccountID [20]byte

// Memo is one entry of the Memos array. Aegis populates only Data, with the
// keccak256 of the on-chain request id — that is what links an XRPL transaction
// to a Flare request in an FDC proof. Type and Format exist because real
// transactions carry them and the serialiser is validated against real blobs.
type Memo struct {
	Type   []byte
	Data   []byte
	Format []byte
}

// Payment is an XRPL Payment in the fields this system needs.
//
// Amounts are drops throughout — 1 XRP is 1,000,000 drops — and never touch a
// float.
type Payment struct {
	Account            AccountID
	Destination        AccountID
	AmountDrops        uint64
	FeeDrops           uint64
	Sequence           uint32
	LastLedgerSequence uint32

	// DestinationTag is omitted from the serialisation when zero. Serialising a
	// zero tag produces a different hash than omitting it, and omitting is what
	// XRPL does.
	DestinationTag uint32

	// Flags is omitted when zero. Aegis always sets TfFullyCanonicalSig, so in
	// this system it is always present; the zero case exists because the
	// reference transactions used to validate this package do not set it.
	Flags uint32

	SigningPubKey []byte
	TxnSignature  []byte
	Memos         []Memo

	// Signers carries the k-of-n signatures of a multi-signed transaction. It is
	// empty for the single-signed path, and when it is populated SigningPubKey
	// is serialised as the empty blob — that emptiness is what tells XRPL to
	// authorise the transaction against the account's SignerList rather than
	// against its master key.
	Signers []Signer
}

// txForm selects which of the signature-carrying fields a serialisation
// includes. The field set is otherwise identical in all four, which is why the
// forms are a parameter rather than four separate functions that could drift.
type txForm uint8

const (
	// formForSigning is what a single signer hashes: SigningPubKey present,
	// TxnSignature absent.
	formForSigning txForm = iota

	// formSigned is a complete single-signed transaction.
	formSigned

	// formForMultiSigning is what each of n signers hashes. SigningPubKey is the
	// empty blob and no signature appears, so every signer hashes identical
	// bytes and only the AccountID appended after them differs.
	formForMultiSigning

	// formMultiSigned is a complete multi-signed transaction: empty
	// SigningPubKey, no TxnSignature, and the Signers array.
	formMultiSigned
)

// multiSigned reports whether the form authorises against a SignerList.
func (f txForm) multiSigned() bool {
	return f == formForMultiSigning || f == formMultiSigned
}

var (
	// ErrNoLastLedgerSequence rejects a transaction that could never be proven
	// non-existent. Without a LastLedgerSequence a stuck payment can neither
	// settle nor be shown not to have settled, which wedges the treasury.
	ErrNoLastLedgerSequence = errors.New("xrpl: LastLedgerSequence is required")

	// ErrNoAmount rejects a zero-value payment.
	ErrNoAmount = errors.New("xrpl: amount must be greater than zero")

	// ErrNoFee rejects a zero fee, which XRPL will not accept.
	ErrNoFee = errors.New("xrpl: fee must be greater than zero")

	// ErrNoSigningPubKey rejects serialisation without a signing key.
	ErrNoSigningPubKey = errors.New("xrpl: SigningPubKey is required")
)

// Validate checks the invariants that must hold before anything is signed.
func (p *Payment) Validate() error {
	return p.validate(false)
}

// validate checks the invariants for one signing mode.
//
// A multi-signed transaction carries no SigningPubKey of its own — the empty
// blob is what selects the SignerList — so that requirement is dropped there
// and nowhere else.
func (p *Payment) validate(multi bool) error {
	if p.LastLedgerSequence == 0 {
		return ErrNoLastLedgerSequence
	}
	if p.AmountDrops == 0 {
		return ErrNoAmount
	}
	if p.FeeDrops == 0 {
		return ErrNoFee
	}
	if !multi && len(p.SigningPubKey) == 0 {
		return ErrNoSigningPubKey
	}
	return nil
}

// Serialize returns the canonical binary form.
//
// When includeSignature is false the TxnSignature field is left out, which is
// the form that gets hashed and signed. Every other field, including
// SigningPubKey, is present in both forms.
func (p *Payment) Serialize(includeSignature bool) ([]byte, error) {
	if includeSignature {
		return p.serialize(formSigned)
	}
	return p.serialize(formForSigning)
}

// serialize builds the canonical binary form for one signing mode.
func (p *Payment) serialize(form txForm) ([]byte, error) {
	fields := []field{
		uint16Field(fieldTransactionType, txTypePayment),
		uint32Field(fieldSequence, p.Sequence),
		uint32Field(fieldLastLedgerSequence, p.LastLedgerSequence),
	}

	if p.Flags != 0 {
		fields = append(fields, uint32Field(fieldFlags, p.Flags))
	}
	if p.DestinationTag != 0 {
		fields = append(fields, uint32Field(fieldDestinationTag, p.DestinationTag))
	}

	amount, err := xrpAmountField(fieldAmount, p.AmountDrops)
	if err != nil {
		return nil, fmt.Errorf("serialising Amount: %w", err)
	}
	fee, err := xrpAmountField(fieldFee, p.FeeDrops)
	if err != nil {
		return nil, fmt.Errorf("serialising Fee: %w", err)
	}
	fields = append(fields, amount, fee)

	// A multi-signed transaction still carries SigningPubKey, as the empty blob.
	// Omitting the field entirely is a different serialisation and a different
	// hash, and XRPL reads the emptiness — not the absence — as "authorise this
	// against the SignerList".
	signingKey := p.SigningPubKey
	if form.multiSigned() {
		signingKey = nil
	}
	pubKey, err := blobField(typeBlob, fieldSigningPubKey, signingKey)
	if err != nil {
		return nil, err
	}
	fields = append(fields, pubKey)

	if form == formSigned {
		sig, err := blobField(typeBlob, fieldTxnSignature, p.TxnSignature)
		if err != nil {
			return nil, err
		}
		fields = append(fields, sig)
	}

	account, err := blobField(typeAccountID, fieldAccount, p.Account[:])
	if err != nil {
		return nil, err
	}
	destination, err := blobField(typeAccountID, fieldDestination, p.Destination[:])
	if err != nil {
		return nil, err
	}
	fields = append(fields, account, destination)

	if len(p.Memos) > 0 {
		memos, err := serializeMemos(p.Memos)
		if err != nil {
			return nil, err
		}
		fields = append(fields, memos)
	}

	if form == formMultiSigned {
		signers, err := serializeSigners(p.Signers)
		if err != nil {
			return nil, err
		}
		fields = append(fields, signers)
	}

	// Sorting here rather than relying on the order above is the point: the
	// canonical order is a property of the codes, not of this function.
	sortFields(fields)

	var out []byte
	for _, f := range fields {
		out = append(out, f.encode()...)
	}
	return out, nil
}

// serializeMemos builds the Memos STArray.
//
// Each entry is an STObject terminated by an object end marker, and the array
// itself by an array end marker.
func serializeMemos(memos []Memo) (field, error) {
	var payload []byte
	for i, m := range memos {
		if len(m.Type) == 0 && len(m.Data) == 0 && len(m.Format) == 0 {
			return field{}, fmt.Errorf("memo %d is empty", i)
		}

		var inner []field
		if len(m.Type) > 0 {
			f, err := blobField(typeBlob, fieldMemoType, m.Type)
			if err != nil {
				return field{}, err
			}
			inner = append(inner, f)
		}
		if len(m.Data) > 0 {
			f, err := blobField(typeBlob, fieldMemoData, m.Data)
			if err != nil {
				return field{}, err
			}
			inner = append(inner, f)
		}
		if len(m.Format) > 0 {
			f, err := blobField(typeBlob, fieldMemoFormat, m.Format)
			if err != nil {
				return field{}, err
			}
			inner = append(inner, f)
		}
		sortFields(inner)

		memoHeader := field{typeCode: typeSTObject, fieldCode: fieldMemo}
		payload = append(payload, memoHeader.header()...)
		for _, f := range inner {
			payload = append(payload, f.encode()...)
		}
		endObject := field{typeCode: typeSTObject, fieldCode: fieldObjectEndMarker}
		payload = append(payload, endObject.header()...)
	}

	endArray := field{typeCode: typeSTArray, fieldCode: fieldArrayEndMarker}
	payload = append(payload, endArray.header()...)

	return field{typeCode: typeSTArray, fieldCode: fieldMemos, payload: payload}, nil
}

File: submitter/src/serialize.ts
/**
 * XRPL transaction serialisation, for assembling a k-of-n signature.
 *
 * Assembly needs no secret and grants no authority: every share already covers
 * the whole transaction, so an assembler that reorders, drops or forges one
 * produces a blob XRPL rejects rather than a payment nobody authorised. That is
 * what lets it run out here rather than in an enclave — and why anyone can run
 * a submitter.
 *
 * This is a third implementation of the same format, after the enclave's and
 * XRPL's own. It is validated against blobs taken off XRPL rather than against
 * the other two, so agreement between them means something.
 */

import { createHash } from "node:crypto";

/** Field type codes, from the XRPL binary format definitions. */
const TYPE_UINT16 = 1;
const TYPE_UINT32 = 2;
const TYPE_AMOUNT = 6;
const TYPE_BLOB = 7;
const TYPE_ACCOUNT_ID = 8;
const TYPE_ST_OBJECT = 14;
const TYPE_ST_ARRAY = 15;

/** Field codes within each type. */
const FIELD_TRANSACTION_TYPE = 2;
const FIELD_FLAGS = 2;
const FIELD_SEQUENCE = 4;
const FIELD_DESTINATION_TAG = 14;
const FIELD_LAST_LEDGER_SEQUENCE = 27;
const FIELD_AMOUNT = 1;
const FIELD_FEE = 8;
const FIELD_SIGNING_PUB_KEY = 3;
const FIELD_TXN_SIGNATURE = 4;
const FIELD_MEMO_TYPE = 12;
const FIELD_MEMO_DATA = 13;
const FIELD_MEMO_FORMAT = 14;
const FIELD_ACCOUNT = 1;
const FIELD_DESTINATION = 3;
const FIELD_SIGNERS = 3;
const FIELD_MEMOS = 9;
const FIELD_MEMO = 10;
const FIELD_SIGNER = 16;
const FIELD_END_MARKER = 1;

const TX_TYPE_PAYMENT = 0;

/** Forbids the malleable signature form. Every Aegis transaction sets it. */
export const TF_FULLY_CANONICAL_SIG = 0x80000000;

/** Transaction id prefix, "TXN\0". */
const PREFIX_TRANSACTION_ID = Buffer.from([0x54, 0x58, 0x4e, 0x00]);

/** The largest blob the variable-length prefix can express. */
const MAX_VL_LENGTH = 918744;

/** The most signatures XRPL accepts on one transaction. */
export const MAX_SIGNERS = 32;

/** One memo. Aegis populates only `data`, with the on-chain request reference. */
export interface Memo {
  readonly type?: Buffer;
  readonly data?: Buffer;
  readonly format?: Buffer;
}

/** One enclave's share of a multi-signed transaction. */
export interface Signer {
  /** The signer's own 20-byte AccountID. */
  readonly account: Buffer;
  readonly signingPubKey: Buffer;
  readonly txnSignature: Buffer;
}

/** An XRPL Payment in the fields this system produces. */
export interface Payment {
  readonly account: Buffer;
  readonly destination: Buffer;
  readonly amountDrops: bigint;
  readonly feeDrops: bigint;
  readonly sequence: number;
  readonly lastLedgerSequence: number;
  readonly destinationTag?: number;
  readonly flags?: number;
  readonly memos?: readonly Memo[];
  /** Present on the single-signed form only; a multi-signed one carries none. */
  readonly signingPubKey?: Buffer;
  readonly txnSignature?: Buffer;
}

interface Field {
  readonly typeCode: number;
  readonly fieldCode: number;
  readonly payload: Buffer;
}

/**
 * Encodes a field id.
 *
 * The high nibble carries the type code and the low nibble the field code, but
 * only when each fits in four bits; anything larger moves to its own trailing
 * byte and leaves its nibble zero. `LastLedgerSequence` (27) and `Signer` (16)
 * are the fields here that take the two-byte form.
 */
function header(typeCode: number, fieldCode: number): Buffer {
  const high = typeCode < 16 ? typeCode << 4 : 0;
  const low = fieldCode < 16 ? fieldCode : 0;

  const out = [high | low];
  if (typeCode >= 16) out.push(typeCode);
  if (fieldCode >= 16) out.push(fieldCode);
  return Buffer.from(out);
}

function encode(field: Field): Buffer {
  return Buffer.concat([header(field.typeCode, field.fieldCode), field.payload]);
}

/**
 * Orders by (typeCode, fieldCode). This is the canonical order — not
 * alphabetical by name, and not the order the fields were built in.
 */
function sortFields(fields: Field[]): Field[] {
  return [...fields].sort((a, b) => (a.typeCode !== b.typeCode ? a.typeCode - b.typeCode : a.fieldCode - b.fieldCode));
}

/** Encodes a variable-length prefix. Three ranges, so a short blob costs a byte. */
function vlPrefix(n: number): Buffer {
  if (n < 0) throw new Error(`negative length ${n}`);
  if (n <= 192) return Buffer.from([n]);
  if (n <= 12480) {
    const v = n - 193;
    return Buffer.from([193 + Math.floor(v / 256), v % 256]);
  }
  if (n <= MAX_VL_LENGTH) {
    const v = n - 12481;
    return Buffer.from([241 + Math.floor(v / 65536), (v >> 8) % 256, v % 256]);
  }
  throw new Error(`blob of ${n} bytes exceeds the maximum ${MAX_VL_LENGTH}`);
}

function blobField(typeCode: number, fieldCode: number, data: Buffer): Field {
  return { typeCode, fieldCode, payload: Buffer.concat([vlPrefix(data.length), data]) };
}

function uint16Field(fieldCode: number, value: number): Field {
  const payload = Buffer.alloc(2);
  payload.writeUInt16BE(value);
  return { typeCode: TYPE_UINT16, fieldCode, payload };
}

function uint32Field(fieldCode: number, value: number): Field {
  const payload = Buffer.alloc(4);
  payload.writeUInt32BE(value >>> 0);
  return { typeCode: TYPE_UINT32, fieldCode, payload };
}

/** The total XRP supply in drops — 100 billion XRP. */
const MAX_DROPS = 100_000_000_000n * 1_000_000n;

/**
 * Builds an XRP amount. Bit 62 marks the value as XRP rather than an issued
 * currency; bit 63 is the sign bit, left clear because an amount is never
 * negative.
 */
function xrpAmountField(fieldCode: number, drops: bigint): Field {
  if (drops < 0n) throw new Error(`negative amount ${drops}`);
  if (drops > MAX_DROPS) throw new Error(`amount ${drops} drops exceeds the total XRP supply`);
  const payload = Buffer.alloc(8);
  payload.writeBigUInt64BE(drops | 0x4000000000000000n);
  return { typeCode: TYPE_AMOUNT, fieldCode, payload };
}

function requireAccountId(value: Buffer, what: string): Buffer {
  if (value.length !== 20) throw new Error(`${what} must be a 20-byte AccountID, got ${value.length}`);
  if (value.every((byte) => byte === 0)) throw new Error(`${what} is zero`);
  return value;
}

/** Builds the Memos STArray. */
function memosField(memos: readonly Memo[]): Field {
  const parts: Buffer[] = [];

  memos.forEach((memo, index) => {
    const inner: Field[] = [];
    if (memo.type?.length) inner.push(blobField(TYPE_BLOB, FIELD_MEMO_TYPE, memo.type));
    if (memo.data?.length) inner.push(blobField(TYPE_BLOB, FIELD_MEMO_DATA, memo.data));
    if (memo.format?.length) inner.push(blobField(TYPE_BLOB, FIELD_MEMO_FORMAT, memo.format));
    if (inner.length === 0) throw new Error(`memo ${index} is empty`);

    parts.push(header(TYPE_ST_OBJECT, FIELD_MEMO));
    for (const field of sortFields(inner)) parts.push(encode(field));
    parts.push(header(TYPE_ST_OBJECT, FIELD_END_MARKER));
  });

  parts.push(header(TYPE_ST_ARRAY, FIELD_END_MARKER));
  return { typeCode: TYPE_ST_ARRAY, fieldCode: FIELD_MEMOS, payload: Buffer.concat(parts) };
}

/** Builds the Signers STArray. */
function signersField(signers: readonly Signer[]): Field {
  const parts: Buffer[] = [];

  signers.forEach((signer, index) => {
    if (signer.signingPubKey.length === 0) throw new Error(`signer ${index} has no public key`);
    if (signer.txnSignature.length === 0) throw new Error(`signer ${index} has no signature`);
    requireAccountId(signer.account, `signer ${index}`);

    const inner = sortFields([
      blobField(TYPE_BLOB, FIELD_SIGNING_PUB_KEY, signer.signingPubKey),
      blobField(TYPE_BLOB, FIELD_TXN_SIGNATURE, signer.txnSignature),
      blobField(TYPE_ACCOUNT_ID, FIELD_ACCOUNT, signer.account),
    ]);

    parts.push(header(TYPE_ST_OBJECT, FIELD_SIGNER));
    for (const field of inner) parts.push(encode(field));
    parts.push(header(TYPE_ST_OBJECT, FIELD_END_MARKER));
  });

  parts.push(header(TYPE_ST_ARRAY, FIELD_END_MARKER));
  return { typeCode: TYPE_ST_ARRAY, fieldCode: FIELD_SIGNERS, payload: Buffer.concat(parts) };
}

/**
 * Puts the shares in the order XRPL requires, refusing any set it would reject.
 *
 * Sorting here rather than trusting the caller makes the order a property of
 * the format instead of of whoever collected the shares — and the contract
 * emits them in arrival order, which is not it.
 */
export function canonicalSigners(signers: readonly Signer[], account: Buffer): Signer[] {
  if (signers.length === 0) throw new Error("a multi-signed transaction needs at least one signer");
  if (signers.length > MAX_SIGNERS) throw new Error(`a transaction accepts at most ${MAX_SIGNERS} signers`);

  const ordered = [...signers].sort((a, b) => Buffer.compare(a.account, b.account));

  for (const signer of ordered) {
    requireAccountId(signer.account, "signer");
    if (signer.account.equals(account)) throw new Error("the sending account cannot be one of its own signers");
  }
  for (let i = 1; i < ordered.length; i += 1) {
    // Non-null: the loop bounds guarantee both indices exist.
    if (ordered[i]!.account.equals(ordered[i - 1]!.account)) throw new Error("the same account signed twice");
  }

  return ordered;
}

/** Which signature-carrying fields a serialisation includes. */
export type PaymentForm = "signed" | "multiSigned";

/**
 * Serialises a Payment.
 *
 * A multi-signed transaction still carries SigningPubKey, as the empty blob.
 * Omitting the field entirely is a different serialisation and a different
 * hash, and XRPL reads the emptiness — not the absence — as "authorise this
 * against the SignerList".
 */
export function serializePayment(payment: Payment, form: PaymentForm, signers: readonly Signer[] = []): Buffer {
  const fields: Field[] = [
    uint16Field(FIELD_TRANSACTION_TYPE, TX_TYPE_PAYMENT),
    uint32Field(FIELD_SEQUENCE, payment.sequence),
    uint32Field(FIELD_LAST_LEDGER_SEQUENCE, payment.lastLedgerSequence),
  ];

  if (payment.flags) fields.push(uint32Field(FIELD_FLAGS, payment.flags));
  // Omitted entirely when zero: serialising a zero tag produces a different
  // hash than omitting it, and omitting is what XRPL does.
  if (payment.destinationTag) fields.push(uint32Field(FIELD_DESTINATION_TAG, payment.destinationTag));

  fields.push(xrpAmountField(FIELD_AMOUNT, payment.amountDrops));
  fields.push(xrpAmountField(FIELD_FEE, payment.feeDrops));

  const signingPubKey = form === "multiSigned" ? Buffer.alloc(0) : (payment.signingPubKey ?? Buffer.alloc(0));
  fields.push(blobField(TYPE_BLOB, FIELD_SIGNING_PUB_KEY, signingPubKey));

  if (form === "signed") {
    if (!payment.txnSignature?.length) throw new Error("a signed transaction needs a TxnSignature");
    fields.push(blobField(TYPE_BLOB, FIELD_TXN_SIGNATURE, payment.txnSignature));
  }

  fields.push(blobField(TYPE_ACCOUNT_ID, FIELD_ACCOUNT, requireAccountId(payment.account, "account")));
  fields.push(blobField(TYPE_ACCOUNT_ID, FIELD_DESTINATION, requireAccountId(payment.destination, "destination")));

  if (payment.memos?.length) fields.push(memosField(payment.memos));
  if (form === "multiSigned") fields.push(signersField(signers));

  return Buffer.concat(sortFields(fields).map(encode));
}

/** SHA-512Half — SHA-512 then the first 32 bytes. Not SHA-256. */
function sha512Half(prefix: Buffer, data: Buffer): Buffer {
  return createHash("sha512")
    .update(prefix)
    .update(data)
    .digest()
    .subarray(0, 32);
}

/** The XRPL transaction id of a signed blob. */
export function transactionId(signedBlob: Buffer): Buffer {
  return sha512Half(PREFIX_TRANSACTION_ID, signedBlob);
}

/** A blob ready to submit, with the id XRPL will know it by. */
export interface AssembledTransaction {
  readonly signedBlob: Buffer;
  readonly txHash: Buffer;
}

/** Combines collected shares into a submittable multi-signed transaction. */
export function assembleMultisigned(payment: Payment, signers: readonly Signer[]): AssembledTransaction {
  if (payment.lastLedgerSequence === 0) {
    // A transaction without an expiry can never be proven non-existent, which
    // breaks the only path that releases a stuck payment's window spend.
    throw new Error("LastLedgerSequence is required");
  }
  if (payment.amountDrops <= 0n) throw new Error("amount must be greater than zero");
  if (payment.feeDrops <= 0n) throw new Error("fee must be greater than zero");

  const ordered = canonicalSigners(signers, payment.account);
  const signedBlob = serializePayment(payment, "multiSigned", ordered);
  return { signedBlob, txHash: transactionId(signedBlob) };
}
